sssd_test_framework/utils/ipa_ca.py

```diff
@@ -0,0 +1,276 @@
+    "superseded",
+    "cessation_of_operation",
+    "certificate_hold",
+    "remove_from_crl",
+    "privilege_withdrawn",
+    "aa_compromise",
+]
+
+
+class IPACertificateAuthority:
+    """
+    Provides FreeIPA Certificate Authority operations.
+    Supports issuing, revoking, placing/removing hold, and querying certificate info.
+    """
+
+    def __init__(self, host: MultihostHost, fs: LinuxFileSystem) -> None:
+        """
+        Initialize IPA CA handler for a remote host.
+
+        :param host: Remote test host
+        :type host: MultihostHost
+        :param fs: Filesystem helper for remote file operations
+        :type fs: LinuxFileSystem
+        """
+        self.host = host
+        self.fs = fs
+        self.temp_dir = f"/tmp/ipa_test_certs_{os.getpid()}_{uuid.uuid4().hex}"
+        self.fs.mkdir_p(self.temp_dir, mode="700")
+
+    def __del__(self) -> None:
+        """
+        Clean up temporary files and directories on object deletion.
+        """
+        self.teardown()
+
+    def teardown(self) -> None:
+        """
+        Remove temporary files/directories created by this handler.
+        """
+        try:
+            self.fs.rm(self.temp_dir)
+        except Exception as e:
+            self.host.logger.warning(f"Failed to remove temporary directory {self.temp_dir}: {e}")
+
+    def request(
+        self,
+        principal: str,
+        subject: Optional[str] = None,
+        validity_days: int = 365,
+        add_service: bool = False,
+        key_size: int = 2048,
+    ) -> Tuple[str, str, str]:
+        """
+        Issue a certificate for a Kerberos principal from the IPA CA.
+
+        If subject is None, derive a subject from the principal.
+
+        :param principal: Kerberos principal, e.g. 'HTTP/host.example.com'
+        :type principal: str
+        :param subject: X.509 subject string, e.g. '/CN=host.example.com'
+        :type subject: Optional[str]
+        :param validity_days: Certificate lifetime in days
+        :type validity_days: int
+        :param add_service: Whether to add the service principal if missing
+        :type add_service: bool
+        :param key_size: RSA private key size in bits
+        :type key_size: int
+        """
+        base = re.sub(r"[^a-zA-Z0-9._-]", "_", principal)
+        key_path = os.path.join(self.temp_dir, f"{base}.key")
+        csr_path = os.path.join(self.temp_dir, f"{base}.csr")
+        cert_path = os.path.join(self.temp_dir, f"{base}.crt")
+
+        if subject is None:
+            if "/" in principal:
+                hostname = principal.split("/", 1)[1].split("@")[0]
+            else:
+                hostname = principal.split("@")[0]
+
+            if not hostname:
+                raise ValueError(f"Cannot derive subject from principal '{principal}'")
+            subject = f"/CN={hostname}"
+
+        if add_service:
+            self.host.conn.run(f"ipa service-add {shlex.quote(principal)}", raise_on_error=False)
+
```

Comment:
You need to add _generate_cert() before the ipa cert-request.   something like this I think:
```
self._generate_csr(key_path, csr_path, subject)
```